Add `onlyAnnotatedFor` flag

Ao-senXiong opened the pull request:
Fixes #1272

Ao-senXiong commented:
Additional pointer to this Java 8 failure. I have not read all of it yet.

wmdietl commented on checker/src/test/java/org/checkerframework/checker/test/junit/NullnessOnlyAnnotatedForTest.java:
```diff
@@ -0,0 +1,34 @@
+import java.io.File;
+import java.util.List;
+
+/** JUnit tests for the Nullness checker. */
```
Add some more explaining what is special about this test.

wmdietl commented on checker/src/test/java/org/checkerframework/checker/test/junit/NullnessOnlyAnnotatedForTest.java:
```diff
@@ -0,0 +1,34 @@
+public class NullnessOnlyAnnotatedForTest extends CheckerFrameworkPerDirectoryTest {
+
+    /**
+     * Create a NullnessNullMarkedTest.
```
Wrong name.

wmdietl commented on docs/CHANGELOG.md:
```diff
@@ -3,6 +3,9 @@ Version 3.49.3-eisop2 (June ??, 2025)
 
 **User-visible changes:**
 
+The new command-line argument `-AonlyAnnotatedFor=true` suppress all type-checking warnings outside the scope of any
+relevant @AnnotatedFor annotation.
```
```suggestion
The new command-line argument `-AonlyAnnotatedFor=true` suppresses all type-checking warnings outside the scope of
relevant @AnnotatedFor annotation.
```

wmdietl commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -2779,7 +2784,7 @@ public boolean shouldSuppressWarnings(@Nullable TreePath path, String errKey) {
         }
 
-        if (useConservativeDefault("source")) {
+        if (useConservativeDefault("source") || this.getBooleanOption("onlyAnnotatedFor", false)) {
```
We should look into the performance of these `only` and `skip` options - currently, I think we perform all the work and then in the end just don't show the errors. We could be faster by not doing work on e.g. a method if we know it's not going to be output.

Ao-senXiong replied:
Opened #1301 for future reference.

wmdietl commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -2985,11 +2999,14 @@ protected boolean messageKeyMatches(
     private boolean isAnnotatedForThisCheckerOrUpstreamChecker(@Nullable Element elt) {
-        if (elt == null || !useConservativeDefault("source")) {
+        if (elt == null
+                || (!useConservativeDefault("source") && !issueErrorsForOnlyAnnotatedForScope)) {
```
This method supposedly answers the question "isAnnotatedForThisCheckerOrUpstreamChecker". Why does it matter whether we use conservative defaults or if errors in an annotatedfor scope should be suppressed?

wmdietl commented:
> Performance perspective: why we did not cache the element is annotatedfor like we did in QualifierDefaults?
> 
> https://github.com/eisop/checker-framework/blob/82d9275ac58caa1fd9183fad2cf3be278fb0d4d7/framework/src/main/java/org/checkerframework/framework/util/defaults/QualifierDefaults.java#L124

What do you mean with this? Should this PR add more caching? Can the two places use the same cache?

wmdietl commented on checker/src/test/java/org/checkerframework/checker/test/junit/NullnessOnlyAnnotatedForTest.java:
```diff
@@ -0,0 +1,37 @@
+                testFiles,
+                org.checkerframework.checker.nullness.NullnessChecker.class,
+                "nullness",
+                "-AonlyAnnotatedFor");
```
Do you expect many tests that will use this flag?
If it's just this one or maybe another one, it's easier to just add a jtreg text under `checker/jtreg/nullness`.

wmdietl commented on docs/manual/warnings.tex:
```diff
@@ -692,6 +692,13 @@
 
+\sectionAndLabel{\code{-AonlyAnnotatedFor} command-line option}{aonlyAnnotatedFor}
+
+You can suppress all errors and warnings for code outside of \code{@AnnotatedFor} by applying this command-line option.
```
Somewhere it should say that the `AnnotatedFor` needs to list that checker's name.
In your test, maybe add something like this:

````
    @AnnotatedFor("regex")
    class A {
        // No expected error, because code is not annotated for nullness.
        Object o = null;
    }
````

(A similar comment for B would be nice.)

wmdietl reviewed (changes requested):
Thanks! Two small comments.
@thisisalexandercook Can you try this and review the change next?

thisisalexandercook commented on docs/manual/annotating-libraries.tex:
```diff
@@ -384,6 +384,8 @@
 conservative defaults
 (see Section~\ref{defaults-classfile}) for any type use with no explicit
 user-written annotation, \emph{and} the checker issues no warnings.
+\code{-AonlyAnnotatedFor} is an option that only suppresses warnings outside of the \<@AnnotatedFor> annotation,
```
```suggestion
The command-line argument \code{-AonlyAnnotatedFor} can be used to  suppress warnings and errors outside of the scope of an \<@AnnotatedFor> annotation,
```

thisisalexandercook replied:
Tested latest commit against [Spring-Framework](https://github.com/spring-projects/spring-framework) with expected results.

wmdietl commented on checker/jtreg/nullness/onlyannotatedfor/NotAnnotatedForNoError.java:
```diff
@@ -0,0 +1,28 @@
+/*
+ * @test
+ *
+ * @summary Test for command-line option -AonlyAnnotatedFor suppresses warnings for code is not annotated for the corresponding checker.
```
```suggestion
 * @summary Test that command-line option -AonlyAnnotatedFor suppresses warnings for code that is not annotated for the corresponding checker.
```

wmdietl commented on docs/CHANGELOG.md:
```diff
@@ -3,6 +3,11 @@ Version 3.49.5-eisop1 (July ??, 2025)
 
 **User-visible changes:**
 
+The new command-line option `-AonlyAnnotatedFor` suppresses all type-checking errors and warnings outside the scope of
+`@AnnotatedFor` annotation.
+Note that the `@AnnotatedFor` annotation must include the checker's name to enable warnings from the checker.
+For example, `@AnnotatedFor("nullness")` for the Nullness Checker.
+
```
```suggestion
The new command-line option `-AonlyAnnotatedFor` suppresses all type-checking errors and warnings outside the scope of
a corresponding `@AnnotatedFor` annotation.
Note that the `@AnnotatedFor` annotation must include the checker's name to enable warnings from that checker.
For example, use `@AnnotatedFor("nullness")` for the Nullness Checker.

```

wmdietl commented on docs/manual/annotating-libraries.tex:
```diff
@@ -384,6 +384,8 @@
 (see Section~\ref{defaults-classfile}) for any type use with no explicit
 user-written annotation, \emph{and} the checker issues no warnings.
+The command-line argument \code{-AonlyAnnotatedFor} can be used to  suppress warnings and errors outside of the scope of an \<@AnnotatedFor> annotation,
+but does not change the default qualifier for source code (See Section~\ref{aonlyAnnotatedFor}).
```
```suggestion
but does not change the default qualifier for source code (See Section~\ref{AonlyAnnotatedFor}).
```

wmdietl commented on docs/manual/introduction.tex:
```diff
@@ -591,6 +591,9 @@
   Section~\ref{askipfiles}.
+\item \<-AonlyAnnotatedFor>
+  Suppress all errors and warnings for code outside the scope of @AnnotatedFor.
+  See Section~\ref{aonlyAnnotatedFor}.
```
```suggestion
  See Section~\ref{AonlyAnnotatedFor}.
```

Ao-senXiong replied:
Maybe let's use `aonlyannotatedfor`, which is consistent with others except for `AuseConservativeDefaultsForUncheckedCodesource`. In fact, I just did a search, `AuseConservativeDefaultsForUncheckedCodesource` is not used anywhere.

wmdietl replied:
I see you're doing this in #1352. With all lower case the section label is quite hard to read.
I would keep the label for `AuseConservativeDefaultsForUncheckedCodesource` unchanged and just consistently use lowercase in this PR (if it is easy enough to read).

wmdietl commented on docs/manual/annotating-libraries.tex:
```diff
@@ -384,6 +384,8 @@
 conservative defaults
 (see Section~\ref{defaults-classfile}) for any type use with no explicit
 user-written annotation, \emph{and} the checker issues no warnings.
+The command-line argument \code{-AonlyAnnotatedFor} can be used to  suppress warnings and errors outside of the scope of an \<@AnnotatedFor> annotation,
```
```suggestion
The command-line argument \code{-AonlyAnnotatedFor} can be used to  suppress errors and warnings outside of the scope of an \<@AnnotatedFor> annotation,
```

wmdietl commented on docs/manual/introduction.tex:
```diff
@@ -591,6 +591,9 @@
   --- or everywhere except within given files or directories/folders.  See
   Section~\ref{askipfiles}.
+\item \<-AonlyAnnotatedFor>
+  Suppress all errors and warnings for code outside the scope of @AnnotatedFor.
```
```suggestion
  Suppress all errors and warnings for code outside the scope of \<@AnnotatedFor>.
```

wmdietl commented on docs/manual/warnings.tex:
```diff
@@ -692,6 +692,15 @@
 Another scenario is when a file contains multiple top-level classes, only
 one of which has a name that corresponds to the file name.
 
+\sectionAndLabel{\code{-AonlyAnnotatedFor} command-line option}{aonlyAnnotatedFor}
```
```suggestion
\sectionAndLabel{\code{-AonlyAnnotatedFor} command-line option}{AonlyAnnotatedFor}
```

wmdietl commented on docs/manual/warnings.tex:
```diff
@@ -692,6 +692,15 @@
+\sectionAndLabel{\code{-AonlyAnnotatedFor} command-line option}{aonlyAnnotatedFor}
+
+You can suppress all errors and warnings for code outside of \code{@AnnotatedFor} by applying this command-line option.
+Note that the \code{@AnnotatedFor} annotation must include the checker's name to enable warnings from the checker.
+For example, \code{@AnnotatedFor("nullness")} for the Nullness Checker.
+This flag only suppresses warnings compared to \code{-AuseConservativeDefaultsForUncheckedCode=source},
+which also applies conservative defaults for code outside the scope of \code{@AnnotatedFor} annotation.
+\href{https://github.com/uber/NullAway}{NullAway} has a similar flag \code{-XepOpt:NullAway:OnlyNullMarked=true}
+configures NullAway to only issue errors for code inside the scope of \code{@NullMarked} annotation.
```
```suggestion
Note that the \code{@AnnotatedFor} annotation must include the checker's name to enable warnings from that checker.
For example, use \code{@AnnotatedFor("nullness")} for the Nullness Checker.
This flag only suppresses warnings compared to \code{-AuseConservativeDefaultsForUncheckedCode=source},
which also applies conservative defaults for code outside the scope of an \code{@AnnotatedFor} annotation.
\href{https://github.com/uber/NullAway}{NullAway} has a similar flag: \code{-XepOpt:NullAway:OnlyNullMarked=true}
configures NullAway to only issue errors for code inside the scope of \code{@NullMarked} annotations.
```

wmdietl reviewed (changes requested):
Thanks! Just a few minor suggested tweaks.

Ao-senXiong commented:
>And discuss the difference in behavior between no command-line flags, the existing -AuseConservativeDefaultsForUncheckedCode flag, and the new flag introduced in this PR?
Add it as a new jtreg test with different expected outputs for the different flags.

Added in this PR. 

>(Also add a test with a field, where the conservative flag doesn't do the right thing - it's only conservative for reading, not for writing the field.)

I also added field write in this PR. I feel like this is seperate issue to fix, I can do it in another PR.

> What is not nice, is that the usage of set will still produce a warning.
So we should discuss that and the usage of "optimistic" defaults for unannotated code - or a different mechanism, like using skipUses for uses of unannotated code.

This also sounds like thing should be done in another PR.

> In the changelog and manual we should discuss the soundness trade-offs between the flags.

I will add something later.

wmdietl commented on checker/jtreg/nullness/onlyannotatedfor/AnnotatedForWithUse.java:
```diff
@@ -0,0 +1,27 @@
+ * @test
+ *
+ * @summary Test that difference default applies to unannotated code.
+ * @compile/fail/ref=OnlyAnnotatedForWithUse.out -XDrawDiagnostics -Xlint:unchecked -processor org.checkerframework.checker.nullness.NullnessChecker -AonlyAnnotatedFor AnnotatedForWithUse.java
```
It's a good idea to use the file name of the test driver as a prefix rather than as a suffix of the expected output file - that puts the files next to each other and makes the grouping clear.

wmdietl commented on checker/jtreg/nullness/onlyannotatedfor/AnnotatedForWithUse.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @summary Test that difference default applies to unannotated code.
+ * @compile/fail/ref=OnlyAnnotatedForWithUse.out -XDrawDiagnostics -Xlint:unchecked -processor org.checkerframework.checker.nullness.NullnessChecker -AonlyAnnotatedFor AnnotatedForWithUse.java
+ * @compile/fail/ref=ConservativeDefaultWithUse.out -XDrawDiagnostics -Xlint:unchecked -processor org.checkerframework.checker.nullness.NullnessChecker -AuseConservativeDefaultsForUncheckedCode=source AnnotatedForWithUse.java
```
Should there be a third execution (first or last?) that executes with the default flags?

wmdietl commented on checker/jtreg/nullness/onlyannotatedfor/AnnotatedForWithUse.java:
```diff
@@ -0,0 +1,27 @@
+    @AnnotatedFor("nullness")
+    class AnnotatedUse {
+        void use(Unannotated u) {
+            @NonNull Object obj =  u.o;
```
Some inline comments would be useful to make it reproducible what you want to test.
Maybe introduce some short-cut notation and say what you expect for each:

````
// 1: OK, 2: Err, but want OK, 3: Err
````

or something like that, where 1, 2, 3 are the different configurations. For each, what is the expected behavior and whether it is what we actually want.
E.g. for the field update, conservative defaults don't work correctly.

wmdietl reviewed (commented):
Add references to existing issues or open new issues for the things that are left open.
At least the unsoundness of the new feature should be mentioned in the changelog.

wmdietl commented:
Reminder that we should use this new flag in https://github.com/eisop/jspecify-conformance and the conformance score should improve.

wmdietl commented:
> Failure is simliar to [jspecify/jspecify-reference-checker#220](https://github.com/jspecify/jspecify-reference-checker/pull/220). Instead of timeout, I got unauthorized.

So, from jspecify that PR made a successful release, but something still doesn't work correctly.
In https://github.com/eisop/jspecify-reference-checker/pull/11 I tried a different URL, but that still fails in the CF CI.
After that, I now get:

````
A problem occurred configuring root project 'jspecify-reference-checker'.
> Could not create task ':conformanceTests'.
   > Could not create task ':unzipConformanceTestSuite'.
      > Could not resolve all files for configuration ':conformanceTestSuite'.
         > Could not find org.jspecify.conformance:conformance-tests:0.0.0-SNAPSHOT.
           Searched in the following locations:
             - file:/home/runner/.m2/repository/org/jspecify/conformance/conformance-tests/0.0.0-SNAPSHOT/maven-metadata.xml
             - file:/home/runner/.m2/repository/org/jspecify/conformance/conformance-tests/0.0.0-SNAPSHOT/conformance-tests-0.0.0-SNAPSHOT.pom
             - https://central.sonatype.com/repository/maven-snapshots/org/jspecify/conformance/conformance-tests/0.0.0-SNAPSHOT/maven-metadata.xml
             - https://central.sonatype.com/repository/maven-snapshots/org/jspecify/conformance/conformance-tests/0.0.0-SNAPSHOT/conformance-tests-0.0.0-SNAPSHOT.pom
             - https://repo.maven.apache.org/maven2/org/jspecify/conformance/conformance-tests/0.0.0-SNAPSHOT/maven-metadata.xml
             - https://repo.maven.apache.org/maven2/org/jspecify/conformance/conformance-tests/0.0.0-SNAPSHOT/conformance-tests-0.0.0-SNAPSHOT.pom
           Required by:
               root project :
````

So that URL also doesn't seem correct.
Maybe it takes a day for the release to be visible? Would be rather odd for a snapshot release.

Any suggestions/PRs welcome.

wmdietl commented on checker/jtreg/nullness/onlyannotatedfor/AnnotatedForWithUse.java:
```diff
@@ -0,0 +1,32 @@
+/*
+ * @test
+ *
+ * @summary Test that difference default applies to unannotated code.
```
```suggestion
 * @summary Test different defaults applied to unannotated code.
```

wmdietl commented on docs/CHANGELOG.md:
```diff
@@ -3,6 +3,13 @@ Version 3.49.5-eisop1 (July ??, 2025)
+a corresponding `@AnnotatedFor` annotation.
+Note that the `@AnnotatedFor` annotation must include the checker's name to enable warnings from that checker.
+For example, use `@AnnotatedFor("nullness")` for the Nullness Checker.
+This option unsoundly use CLIMB default and suppresses the warnings outside the scope of a corresponding `@AnnotatedFor` annotation.
+Use `-AuseConservativeDefaultsForUncheckedCode=source` if you want conservative default for source code outside the scope of a corresponding `@AnnotatedFor` annotation.
```
```suggestion
This option unsoundly uses source defaults and suppresses the warnings outside the scope of a corresponding `@AnnotatedFor` annotation.
Use `-AuseConservativeDefaultsForUncheckedCode=source` if you want conservative defaults for source code outside the scope of a corresponding `@AnnotatedFor` annotation.
```

wmdietl commented on docs/manual/annotating-libraries.tex:
```diff
@@ -384,6 +384,8 @@
 (see Section~\ref{defaults-classfile}) for any type use with no explicit
 user-written annotation, \emph{and} the checker issues no warnings.
+The command-line argument \code{-AonlyAnnotatedFor} can be used to  suppress errors and warnings outside of the scope of an \<@AnnotatedFor> annotation,
+but does not change the default qualifier for source code (See Section~\ref{aonlyannotatedfor}).
```
```suggestion
but does not change the default qualifiers for source code (See Section~\ref{aonlyannotatedfor}).
```

wmdietl commented on docs/manual/introduction.tex:
```diff
@@ -591,6 +591,10 @@
   Section~\ref{askipfiles}.
+\item \<-AonlyAnnotatedFor>
+  Suppress all errors and warnings for code outside the scope of \<@AnnotatedFor> and
+  unsoundly use CLIMB default and suppresses the warnings outside the scope of a corresponding \<@AnnotatedFor> annotation.
+  See Section~\ref{aonlyannotatedfor}.
```
```suggestion
  unsoundly use source defaults.
  See Section~\ref{aonlyannotatedfor}.
```

wmdietl commented on docs/manual/introduction.tex:
```diff
@@ -591,6 +591,10 @@
   --- or everywhere except within given files or directories/folders.  See
   Section~\ref{askipfiles}.
+\item \<-AonlyAnnotatedFor>
+  Suppress all errors and warnings for code outside the scope of \<@AnnotatedFor> and
```
```suggestion
  Suppress all errors and warnings for code outside the scope of a corresponding \<@AnnotatedFor> and
```

wmdietl commented on docs/manual/warnings.tex:
```diff
@@ -692,6 +692,15 @@
 
+\sectionAndLabel{\code{-AonlyAnnotatedFor} command-line option}{aonlyannotatedfor}
+
+You can suppress all errors and warnings for code outside of \code{@AnnotatedFor} by applying this command-line option.
```
```suggestion
You can suppress all errors and warnings for code outside of a corresponding \code{@AnnotatedFor} by applying this command-line option.
```

wmdietl commented on docs/manual/warnings.tex:
```diff
@@ -692,6 +692,15 @@
+You can suppress all errors and warnings for code outside of \code{@AnnotatedFor} by applying this command-line option.
+Note that the \code{@AnnotatedFor} annotation must include the checker's name to enable warnings from that checker.
+For example, use \code{@AnnotatedFor("nullness")} for the Nullness Checker.
+This flag only suppresses warnings compared to \code{-AuseConservativeDefaultsForUncheckedCode=source},
```
```suggestion
This flag only suppresses warnings, compared to \code{-AuseConservativeDefaultsForUncheckedCode=source},
```

wmdietl commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -141,6 +141,20 @@
     "onlyFiles",
     "skipDirs", // Obsolete as of 2024-03-15, replaced by "skipFiles".
 
+    // Suppress all errors and warnings for code outside the scope of @AnnotatedFor annotation.
```
```suggestion
    // Suppress all errors and warnings for code outside the scope of a corresponding `@AnnotatedFor` annotation.
```

wmdietl commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -141,6 +141,20 @@
+    // Note that the `@AnnotatedFor` annotation must include the checker's name to enable warnings
+    // from that checker.
+    // For example, use `@AnnotatedFor("nullness")` for the Nullness Checker.
+    // This flag only suppresses warnings compared to
```
```suggestion
    // This flag only suppresses warnings, compared to
```

wmdietl commented on framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:
```diff
@@ -197,7 +211,9 @@
-    // sections in the manual for more details
+    // sections in the manual for more details.
+    // -AonlyAnnotatedFor suppresses warnings for code outside the scope of @AnnotatedFor,
+    // but does not change the default qualifier for source code.
```
```suggestion
    // but does not change the default qualifiers for source code.
```

wmdietl reviewed (changes requested):
Thanks! A few small wording suggestions.
You might need to run reformatting to make CI pass. Please also pull in `master`.
Thanks!

wmdietl reviewed (approved):
Thanks!
